pkg/controllers/updaterun/initialization.go

```diff
@@ -0,0 +1,433 @@
+		}
+		return "", err
+	}
+	// Check if the ClusterResourcePlacement has an external rollout strategy.
+	if crp.Spec.Strategy.Type != placementv1beta1.ExternalRolloutStrategyType {
+		klog.V(2).InfoS("The clusterResourcePlacement does not have an external rollout strategy", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		return "", fmt.Errorf("%w: %s", errInitializedFailed, "parent clusterResourcePlacement does not have an external rollout strategy, current strategy: "+string(crp.Spec.Strategy.Type))
+	}
+	updateRun.Status.ApplyStrategy = crp.Spec.Strategy.ApplyStrategy
+	return crp.Name, nil
+}
+
+// determinePolicySnapshot retrieves the latest policy snapshot associated with the ClusterResourcePlacement,
+// and validates it and records it in the ClusterStagedUpdateRun status.
+func (r *Reconciler) determinePolicySnapshot(
+	ctx context.Context,
+	placementName string,
+	updateRun *placementv1alpha1.ClusterStagedUpdateRun,
+) (*placementv1beta1.ClusterSchedulingPolicySnapshot, int, error) {
+	updateRunRef := klog.KObj(updateRun)
+	// Get the latest policy snapshot.
+	var policySnapshotList placementv1beta1.ClusterSchedulingPolicySnapshotList
+	latestPolicyMatcher := client.MatchingLabels{
+		placementv1beta1.CRPTrackingLabel:      placementName,
+		placementv1beta1.IsLatestSnapshotLabel: "true",
+	}
+	if err := r.Client.List(ctx, &policySnapshotList, latestPolicyMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list the latest policy snapshots", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		// err can be retried.
+		return nil, -1, err
+	}
+	if len(policySnapshotList.Items) != 1 {
+		if len(policySnapshotList.Items) > 1 {
+			err := fmt.Errorf("more than one (%d in actual) latest policy snapshots associated with the clusterResourcePlacement: %s", len(policySnapshotList.Items), placementName)
+			klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Failed to find the latest policy snapshot", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+			// no more retries for this error.
+			return nil, -1, fmt.Errorf("%w: %s", errInitializedFailed, err.Error())
+		}
+		// no latest policy snapshot found.
+		err := fmt.Errorf("no latest policy snapshot associated with the clusterResourcePlacement: %s", placementName)
+		klog.ErrorS(err, "Failed to find the latest policy snapshot", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		// again, no more retries here.
+		return nil, -1, fmt.Errorf("%w: %s", errInitializedFailed, err.Error())
+	}
+
+	latestPolicySnapshot := policySnapshotList.Items[0]
+	updateRun.Status.PolicySnapshotIndexUsed = latestPolicySnapshot.Name
+	// Get the cluster count from the policy snapshot.
+	clusterCount, err := annotations.ExtractNumOfClustersFromPolicySnapshot(&latestPolicySnapshot)
```

Comment:
From the code [here](url), the annotation is only added when the it's "pickN" strategy. This may not be correct. Need verification here.

Reply:
I think you are correct, we should special handle the case that this annotation does not exsit.